### DISCUSSION TITLE
Getting transpose of edge features

### DISCUSSION THREAD

**Discussion by PabloVD** [pyg-team/pytorch_geometric#6291] [Q&A]
I generate a graph using for instance `torch_cluster.radius_graph`, so the edges have shape $(2, N_e)$. Then I build edge features $e_{ij}$ between nodes nodes $i$ and $j$. with shape $(N_e,C)$, being $C$ the number of channels. Note that in general $e_{ij}$  is different from $e_{ji}$.

Given the edge features $e_{ij}$, how can I get the transpose edge feature, $e_{ji}$? Is there any efficient way of doing that in pytorch geometric?

**Comment by wsad1** [comment]
I am a little confused by the question. What do you mean by transpose of edge features? Lets say your graph has a single edge [[1,2]]
and the feature vector for that edge is lets say [0.1, 0.2, 0.3]. The edge transpose would be [[2,1]], what would the transpose of the features be? Do let me know if I misunderstood something.

**Reply by PabloVD** [reply]
Thanks for replying, let me clarify that. 
Imagine that our graph has 2 nodes and node features `x`. Then, building the edges for instance with `edge_index = radius_graph(x, ...)` we have two directional edges, `edge_index = tensor([ [1, 0], [0, 1] ])`.
Now lets create edge features as
```
row, col = edge_index[0], edge_index[1]
edge_feat = x[row]-x[col]
```
In that case, getting the transpose of the edge features would be easy, just `x[col]-x[row]`, and thus $e_{ji} = -e_{ij}$.

But if I update the edge features using some multilayer perceptron, $e_{ij}'=MLP(e_{ij})$, that antisymmetry relation does not hold anymore, and I would like to have access to the transpose of the edge features  $e_{ji}'$, getting the proper permutation of indexes of the edge features.

**Reply by rusty1s** [reply]
Note that `radius_graph` does not necessarily produce undirected graph, so each edge may not have a reverse edge. Currently, there exists no efficient way to query the reverse edge using `edge_index`. You may find some success using `SparseTensor` though (untested):
```python
from torch_sparse import SparseTensor
adj = SparseTensor.from_edge_index(edge_index, edge_attr)
e_ij = adj[i, j]
e_ji = adj[j, i]
```

**Reply by PabloVD** [reply]
Many thanks for your answer. Could you please explain how you define indexes `i` and `j` above? Are equivalent to `row, col = edge_index[0], edge_index[1]`?

**Reply by PabloVD** [reply]
That works! I managed to get the transpose tensor `edge_attr_trans` from a given tensor of edge features `edge_attr`. Not very efficient but works, many thanks @rusty1s !
```python
from torch_sparse import SparseTensor
adj = SparseTensor.from_edge_index(edge_index, edge_attr)

edge_attr_trans = torch.zeros_like(edge_attr, device=device)

for edge in range(edge_index.shape[1]):
    i, j = edge_index[0,edge].item(), edge_index[1,edge].item()
    e_ji = adj[j, i].to_dense()
    edge_attr_trans[edge] = e_ji
```

**Reply by rusty1s** [reply]
Nice, I think you can also use `j` and `i` as vectors, which avoids the need of the for-loop.

**Reply by PabloVD** [reply]
Not sure how. If I try
```python
i, j = edge_index[0], edge_index[1]
e_ji = adj[j, i].to_dense()
```
then, the resulting array has shape $(N_e, N_e, C)$ rather than the expected $(N_e, C)$ (being $N_e$ the number of edges and $C$ the number of channels).

**Reply by rusty1s** [reply]
Ah, you are right. Sorry about that. It looks like we really don't support this yet efficiently in `SparseTensor` :(
